[KYUUBI #3424] [FEATURE] [AUTHZ] Access privilege checks for namespaces and tables of DatasourceV2

bowenliang123 opened the pull request:
<!--
Thanks for sending a pull request!

Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://kyuubi.readthedocs.io/en/latest/community/CONTRIBUTING.html
  2. If the PR is related to an issue in https://github.com/apache/incubator-kyuubi/issues, add '[KYUUBI #XXXX]' in your PR title, e.g., '[KYUUBI #XXXX] Your PR title ...'.
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][KYUUBI #XXXX] Your PR title ...'.
-->

### _Why are the changes needed?_
<!--
Please clarify why the changes are needed. For instance,
  1. If you add a feature, you can talk about the use case of it.
  2. If you fix a bug, you can clarify why it is a bug.
-->
close #3424.

Covering the following V2 commands,
- v2Commands
  - CreateNamespace: CREATE DATABASE
  - DropNamespace: DROP DATABASE
  - CommentOnNamespace
  - CreateTable
  - CreateV2Table
  - CreateAsSelect
  - UpdateTable
  - AppendData: Insert Using a SELECT Statement /  VALUES / TABLE
  - OverwriteByExpression: Insert Overwrite
  - OverwritePartitionsDynamic
  - CacheTable
  - DropTable
  - TruncateTable
  - AlterTable
  - MergeIntoTable
  - RepairTable

- v2AlterTableCommands
  - CommentOnTable
  - AddColumns
  - DropColumns
  - RenameColumn
  - AlterColumn
  - ReplaceColumns

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

bowenliang123 commented:
I have change the v2 catalog class for test suite from InMemoryTableCatalog (3.2+) to `org.apache.spark.sql.execution.datasources.v2.jdbc.JDBCTableCatalog`(3.1+) and made all ut passed with leverged differnences between 3.2+ and 3.1 and below.

As for Spark 3.0, I don't have good tools for e2e test as for 3.1+.  V2 catalog seems not so mature in 3.0. I am considering leaving it behind and port back 3.0 in the future..

@hellojinsilei

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:
```diff
@@ -93,6 +101,22 @@ object PrivilegesBuilder {
       }
     }
 
+    def mergeProjectionWithIdentifier(
```
private

bowenliang123 replied:
As method `mergeProjectionWithIdentifier`  inside method `buildQuery`, 'private' modifier is unnecessary and not allowed here.

yaooqinn replied:
sorry, my bad

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:
```diff
@@ -93,6 +101,22 @@ object PrivilegesBuilder {
 
+    def mergeProjectionWithIdentifier(
+        table: Identifier,
+        cols: Seq[Attribute],
```
can be removed

bowenliang123 replied:
removed and reuse plan for column output listing. thx for your code sample.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:
```diff
@@ -93,6 +101,22 @@ object PrivilegesBuilder {
+        table: Identifier,
+        cols: Seq[Attribute],
+        plan: LogicalPlan,
+        actionType: PrivilegeObjectActionType = PrivilegeObjectActionType.OTHER): Unit = {
```
can be removed

bowenliang123 replied:
removed, thx.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:
```diff
@@ -47,6 +48,13 @@ object PrivilegesBuilder {
     PrivilegeObject(TABLE_OR_VIEW, actionType, table.database.orNull, table.table, columns)
   }
 
+  private def tablePrivilegesWithIdentifier(
```
how about v2TablePrivileges and v1...

bowenliang123 replied:
good suggestion.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:
```diff
@@ -93,6 +101,16 @@ object PrivilegesBuilder {
       }
     }
 
+    def mergeProjectionWithIdentifier(table: Identifier, plan: LogicalPlan): Unit = {
```
how about projectV2Table and projectV1Table

bowenliang123 replied:
good. much clearer.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:
```diff
@@ -273,6 +308,18 @@ object PrivilegesBuilder {
 
+      case "CreateNamespace" =>
+        // for Spark 3.1 and below
+        val namespace = getFieldValOption[Seq[String]](plan, "namespace")
```
use the spark version as condition instead of the current way, FYI, SetCatalogAndNamespace as an example

bowenliang123 replied:
applied.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:
```diff
@@ -335,13 +381,20 @@ object PrivilegesBuilder {
-            val db = Some(quote(nameParts.init))
-            outputObjs += tablePrivileges(TableIdentifier(nameParts.last, db))
-          case _ =>
+        val tableIdent = getFieldValOption[Identifier](plan, "tableName")
```
ditto

bowenliang123 replied:
changed to getPlanField

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:
```diff
@@ -411,6 +469,26 @@ object PrivilegesBuilder {
+        val table = getPlanField[AnyRef]("table")
+        val tableIdent = getFieldVal[Option[Identifier]](table, "identifier")
+        outputObjs += tablePrivilegesWithIdentifier(tableIdent.get, actionType = UPDATE)
+      // todo INSERT ?
```
if you add TODO, better to create an issue to track. otherwise, todo is never to be done.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:
```diff
@@ -411,6 +469,26 @@ object PrivilegesBuilder {
         outputObjs += tablePrivileges(table, actionType = actionType)
         buildQuery(getQuery, inputObjs)
 
+      case "AppendData" => // DsV2 insert
```
the comment here is useless.

How about we group v2 commands together with or without a new method

bowenliang123 replied:
I prefer group v2commands in set or enum , but not to introduce new method for handling them. Stay in one method is good enough for now.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/util/AuthZUtils.scala:
```diff
@@ -98,14 +109,14 @@ private[authz] object AuthZUtils {
+  }
+
+  def getTableIdentifierFromIdentifier(id: Identifier): TableIdentifier = {
+    val namespaces = invoke(id, "namespace").asInstanceOf[Array[String]]
```
what not call id.namespace() directly?

bowenliang123 replied:
ok, it's a legacy usage from 1.6.0. let me fix it.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:
```diff
@@ -273,6 +308,18 @@ object PrivilegesBuilder {
           inputObjs += databasePrivileges(db.get)
         }
 
+      case "CreateNamespace" =>
```
the command here is out of the scope of this PR, at least it does match the PR title

bowenliang123 replied:
Change the PR and issue with extended command list for covering both databases/namespaces and tables.

bowenliang123 commented:
With completely refractored v2 commands handling by dedicate privilige building handlers.

1. introduced command privilege handling  framework/pattern  for incoming changing commands from Spark 3.x+ or other Engine / Plugins, eg. Arctics.
2. resue possible building process and added support for least/most spark version , the command type feature grouping commands into `V2CreateTablePlan/V2CreateTablePlan/V2AlterTableCommand` for output checks and `HasQuery` for subquery input checks.
3. prevent unbound command name in string and group in command enums.
4. introducing minimum changes into PrivilegeBuilder as possible, by extended dedicate comand list and builders.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/v2Commands.scala:
```diff
@@ -0,0 +1,346 @@
+  val V2CreateTablePlan, V2WriteCommand, V2AlterTableCommand, V2PartitionCommand = Value
+
+  // custom types
+  val HasQuery, V2DdlTableCommand = Value
```
is V2CreateTablePlan a type of V2DdlTableCommand

bowenliang123 replied:
no, V2DdlTableCommand is customed type here for building outputs for extract target table by same pattern.

bowenliang123 replied:
V2CreateTablePlan is directly from traits of Spark's` v2Command.scala`. The V2DdlTableCommand is a customed cmds solely in Kyuubi Authz. They also differ in detail implements for building outputObjs.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/v2Commands.scala:
```diff
@@ -0,0 +1,346 @@
+      operType: OperationType = QUERY,
+      cmdTypes: Seq[V2CommandType] = Seq(),
+      outputActionType: PrivilegeObjectActionType = PrivilegeObjectActionType.OTHER,
+      enabled: Boolean = true)
```
what does the parameter `enabled` mean？

bowenliang123 replied:
It's meant to control whether enabled this command mapping. Handy for listing unimplemented commands like CreateView and etx. Also handy for debug and customized commands.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/v2Commands.scala:
```diff
@@ -0,0 +1,346 @@
+    }
+
+  case class V2Command(
+      leastVer: String = "",
```
use Option[String]

bowenliang123 replied:
ok, will follow it.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/v2Commands.scala:
```diff
@@ -0,0 +1,346 @@
+      cmdTypes: Seq[V2CommandType] = Seq(),
+      outputActionType: PrivilegeObjectActionType = PrivilegeObjectActionType.OTHER,
+      enabled: Boolean = true)
+    extends super.Val {
```
do we need this?

bowenliang123 replied:
removed.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/v2Commands.scala:
```diff
@@ -0,0 +1,346 @@
+  // with V2AlterTableCommand
+
+  val AlterTable: V2Command = V2Command(
+    operType = ALTERTABLE_ADDCOLS,
+    leastVer = "3.0",
+    mostVer = "3.1",
+    buildOutput = (plan, outputObjs, _, _) => {
+      val table = getFieldVal[Any](plan, "table")
+      val tableIdent = getFieldVal[Option[Identifier]](table, "identifier")
+      if (tableIdent.isDefined) {
+        outputObjs += v2TablePrivileges(tableIdent.get)
+      }
```
can we adjust the order? here and other related places

bowenliang123 replied:
changed field order of V2Command.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/v2Commands.scala:
```diff
@@ -0,0 +1,346 @@
+          ArrayBuffer[PrivilegeObject],
+          Seq[V2CommandType],
+          PrivilegeObjectActionType) => Unit = defaultBuildOutput,
+      operType: OperationType = QUERY,
```
nit:opType or operationType

bowenliang123 replied:
ok, applied.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/v2Commands.scala:
```diff
@@ -0,0 +1,346 @@
+      }
+    }
+
+  case class V2Command(
```
CommandConverter/CommandDetail/CommandInfo etc

bowenliang123 replied:
changed to `CmdPrivilegeBuilder`

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/v2Commands.scala:
```diff
@@ -0,0 +1,343 @@
+    PrivilegeObject(TABLE_OR_VIEW, actionType, quote(table.namespace()), table.name(), columns)
+  }
+
+  // /////////////////////////////////////////////////////////////////////////////////////////////
```
remove this

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/v2Commands.scala:
```diff
@@ -0,0 +1,343 @@
+  type V2CommandType = Value
+
+  // traits' name from Spark v2commands and v2AlterTableCommands
+  val V2CreateTablePlan, V2WriteCommand, V2AlterTableCommand, V2PartitionCommand = Value
```
How about naming them all like HasQuery？ HasQueryField, HasTableNameAs... etc

bowenliang123 replied:
Good Suggestion. Changed all the names from Spark command traits to more abstract identifications as HasChildAsIdentifier, HasQueryAsLogicPlan, HasTableAsIdentifier, HasTableAsIdentifierOption, HasTableNameAsIdentifier.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/v2Commands.scala:
```diff
@@ -0,0 +1,358 @@
+   */
+  object CommandType extends Enumeration {
+    type CommandType = Value
+    val HasChildAsIdentifier, HasQueryAsLogicPlan, HasTableAsIdentifier, HasTableAsIdentifierOption,
```
logical plan

bowenliang123 replied:
fixed. Thanks.

yaooqinn commented on extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/v2Commands.scala:
```diff
@@ -0,0 +1,358 @@
+ */
+object v2Commands extends Enumeration {
+
+  /**
+   * Command types
+   * for hinting privileges building of inputObjs or outputObjs
+   */
```
better to elaborate a bit of the naming rule

bowenliang123 replied:
Fixed with added comment  with naming rule as `HasFieldAsReturnType`.

yaooqinn commented:
thanks, merged to master
